lib/rack/lint.rb

```diff
@@ -485,6 +485,11 @@ def each(*args)
         def close(*args)
           @input.close(*args)
         end
+
+        ## * +set_encoding+ just passes through to the underlying object.
+        def set_encoding(*args)
+          @input.set_encoding(*args) if @input.respond_to?(:set_encoding)
```

Comment:
Is this necessary? Surely this is on the mock request to deal with - i.e. we shouldn't change the interface requirements of the input.

Reply:
I encountered this situation when appropriating `Rack::MockRequest.env_for` to generate the elements for (what are effectively) subrequests within an app, what happened was `rack.input` was an instance of `Rack::Lint::Wrapper::InputWrapper` wrapping an instance of `Puma::NullIO` and neither respond to `#set_encoding`, so it would crash.

I genuinely believe it's more clement behaviour for `Rack::MockRequest.env_for` to test if the input object responds to `#set_encoding` before trying to call it (just like it does for `#size` directly below).

 I'm not married to that pass-through method in `Rack::Lint::Wrapper::InputWrapper` but the motivation there was [Postel's law](https://en.wikipedia.org/wiki/Robustness_principle). In case the underlying input _does_ respond to `#set_encoding`, perform the expected behaviour.

Reply:
What you are asking, is for every implementation of the input body to optionally adopt this interface. But the expectation is, the input should always be binary. So, I'm not sure whether this is a good idea.

Reply:
I am doing nothing more than ask for things not to crash when they don't need to. Prior to this patch, the implementation of `Rack::MockRequest.env_for` does exactly that: crash because it already assumes whatever handed to it responds to `#set_encoding`. Do you have an issue with that too, or just the pass-through method in `Rack::Lint::Wrapper::InputWrapper`?